File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/2552-maximum-sum-of-distinct-subarrays-with-length-k.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        size=0
        start=0
        sumi,maxi=0,0
        hash=dict()
        for i in range(len(nums)):
            if nums[i] not in hash:
                hash[nums[i]]=0
            hash[nums[i]]+=1
            size+=1
            sumi+=nums[i]
            while size>k:
                hash[nums[start]]-=1
                if hash[nums[start]]==0:
                    del hash[nums[start]]
                size-=1
                sumi-=nums[start]
                start+=1
            if k==size and len(hash)==k:
                maxi=max(maxi,sumi)  
        return maxi
                    

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/2552-maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        int start=0,size=0;
        long sum=0,max=0;
        HashMap<Integer,Integer> hash=new HashMap<>();
        for(int i=0;i<nums.length;i++){
            hash.put(nums[i],hash.getOrDefault(nums[i],0)+1);
            size++;
            sum+=nums[i];
            while(size>k){
                hash.put(nums[start],hash.get(nums[start])-1);
                if(hash.get(nums[start])==0){
                    hash.remove(nums[start]);
                }
                size--;
                sum-=nums[start];
                start++;
            }
            if(size==k && hash.size()==k){
                max=Math.max(max, sum);
            }
        }
        return max;
    }
}